File: CPPSolutions/LeetCode/DP/M3144_分割字符频率相等的最少子字符串.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * @author Hyperspace
 * @date 2024/08/28
 * @file M3144_分割字符频率相等的最少子字符串.cpp
 */
class Solution {
public:
    int minimumSubstringsInPartition(string s) {
        int n = s.length();
        int f[n], freq[26];
        for (int i = 0; i < n; i++) f[i] = i + 1;
        auto check = [&](int len, int size) -> bool {
            if (len % size) return false;
            int cnt = len / size;
            for (int v: freq)
                if (v > 0 && v != cnt)
                    return false;

            return true;
        };
        for (int i = 0; i < n; i++) {
            memset(freq, 0, sizeof(freq));
            int size = 0;
            for (int j = i; j >= 0; j--) {
                if (!freq[s[j] - 'a']++) size++;
                if (check(i - j + 1, size)) f[i] = !j ? 1 : min(f[i], f[j - 1] + 1);
            }
        }

        return f[n - 1];
    }
};

File: GolangSolutions/DP/M3144_分割字符频率相等的最少子字符串.go
/*
	@author Hyperspace
	@date 2024/08/28
	@file M3144_分割字符频率相等的最少子字符串.go
*/

package DP

func minimumSubstringsInPartition(s string) int {
	f, freq := make([]int, len(s)), make([]int, 26)
	for i := range f {
		f[i] = i + 1
	}
	var check func(int, int) bool
	check = func(len, size int) bool {
		if len%size != 0 {
			return false
		}
		cnt := len / size
		for _, v := range freq {
			if v > 0 && v != cnt {
				return false
			}
		}

		return true
	}
	for i := range f {
		size := 0
		for j := i; j >= 0; j-- {
			freq[s[j]-'a']++
			if freq[s[j]-'a'] == 1 {
				size++
			}
			if check(i-j+1, size) {
				if j == 0 {
					f[i] = 1
				} else {
					f[i] = min(f[i], f[j-1]+1)
				}
			}
		}
		freq = make([]int, 26)
	}

	return f[len(s)-1]
}
